### PR TITLE
Allow messages to override the linterName (in v2)

### LINKED ISSUE #1485: Allow overriding `linterName` in messages?
In the Language Service Protocol (https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md), diagnostics (`interface Diagnostic`) can specify dynamic sources.

One use case for this might be for a JavaScript language server, where diagnostics can come from both a typing system (Flow, TypeScript) and linters (ESLint); it makes sense to distinguish between them.

When bridging this with the Linter API (as https://github.com/atom/atom-languageclient is doing), this presents a slight issue as it's designed around registering fixed source names. It's certainly possible to dynamically register linters as diagnostics come in (especially with Indie v2), but that's rather heavyweight.

Would it be reasonable to allow for users to override `linterName` in the Message v2 API (from https://github.com/steelbrain/linter/blob/master/docs/types/linter-message-v2.md)? The semantics would be pretty straightforward - if you provide a custom `linterName` then Linter would use it, but otherwise it'd default to the name of the linter.

See: https://github.com/atom/atom-languageclient/issues/47

### PR DESCRIPTION
**Pull request by hansonw**
This PR allows providers to override the displayed `linterName` on a per-message level.
I added a spec as well as a validator+spec.

Released under CC0.
Closes #1485

### CODE REVIEW

File: docs/types/linter-message-v2.md [hunk 1]
```diff
@@ -44,12 +44,13 @@ type Message = {
   description?: string | (() => Promise<string> | string)
   // ^ Markdown long description of the error, accepts callback so you can do
   // http requests etc.
+  linterName?: string,
+  // ^ Optionally override the displayed linter name. (Defaults to provider)
```
Comment (Arcanemagus) [inline docs/types/linter-message-v2.md]: "Defaults to provider name"

File: docs/types/linter-message-v2.md [hunk 2]
```diff
@@ -44,12 +44,13 @@ type Message = {
   description?: string | (() => Promise<string> | string)
   // ^ Markdown long description of the error, accepts callback so you can do
   // http requests etc.
+  linterName?: string,
+  // ^ Optionally override the displayed linter name. Defaults to provider name
```
Comment (Arcanemagus) [inline docs/types/linter-message-v2.md]: Add a `.` on the end, or bring back the `()`, either way works.

**Comment by Arcanemagus** [comment]
Looks like the AppVeyor failures are due to the new async nature of `TextEditor::save()`, putting up a PR fixing the spec for that.

**Comment by hansonw** [comment]
:+1: done

**Comment by Arcanemagus** [comment]
@steelbrain #1494 should be merged before this, or `master` builds will fail.

File: lib/helpers.js [hunk 3]
```diff
@@ -98,7 +98,7 @@ export function normalizeMessages(linterName: string, messages: Array<Message>)
       }
     }
     message.version = 2
-    message.linterName = linterName
+    message.linterName = message.linterName || linterName
```
Comment (steelbrain) [inline lib/helpers.js]: Can we instead do
```
if (!message.linterName) {
  message.linterName = linterName
}
```

Really minor but personal preference :)

**Comment by hansonw** [comment]
Done!

**Comment by steelbrain** [comment]
Perfect 👌 Now a rebase on master/merge from master and it'll be good to land 😃

**Comment by steelbrain** [comment]
Thanks again for working on this 👍

**Comment by steelbrain** [comment]
Released in v2.2.0 🎉

**Pull request merged**